languages/java/oso/src/main/java/com/osohq/oso/Host.java

```diff
@@ -103,37 +91,50 @@ public Long cacheInstance(Object instance, Long id) throws Exceptions.OsoExcepti
     /**
      * Make an instance of a Java class from a {@code Map<String, Object>} of
      * fields.
-     *
-     * @param clsName
-     * @param fields
-     * @param id
      */
-    public Object makeInstance(String clsName, Map fields, long id) throws Exceptions.OsoException {
-        Function<Map, Object> constructor = constructors.get(clsName);
+    public Object makeInstance(String className, List<Object> initargs, long id)
+            throws Exceptions.OsoException
+    {
+        Constructor<?> constructor = constructors.get(className);
+        if (constructor == null) {
+            // Try to find a constructor applicable to the supplied arguments.
+            Class<?> cls = classes.get(className);
+            Class<?>[] argTypes = initargs.stream().map(arg -> arg.getClass())
+                .collect(Collectors.toUnmodifiableList())
+                .toArray(new Class[0]);
+            search: for (Constructor<?> c : cls.getConstructors()) {
+                Class<?>[] paramTypes = c.getParameterTypes();
+                if (argTypes.length == paramTypes.length) {
+                    for (int i = 0; i < paramTypes.length; i++) {
+                        if (!paramTypes[i].isAssignableFrom(argTypes[i])) {
+                            continue search;
+                        }
+                    }
+                    constructor = c;
+                    break search;
+                }
+            }
```

Comment:
Is there a reason not to use [`getConstructor()`](https://docs.oracle.com/javase/7/docs/api/java/lang/Class.html#getConstructor(java.lang.Class...))? If it's because of `isAssignableFrom`, why is this necessary?

Reply:
Consider this constructor: `public Http(String hostname, String path, Map<String, String> query)`. This is the right signature for it to have, but `Map` is an interface; no actual argument will ever have `arg.class == Map`. But a `HashMap`, say, `isAssignableFrom` a `Map`, as is any subclass.

The `Class.getMethod` method does all this, but `getConstructor` doesn't. I don't think there's a good reason for this, other than "because Java".

Reply:
ugh wow